### ISSUE TITLE
Release PyGMT v0.14.0

### ISSUE THREAD

**Issue by seisman** [GenericMappingTools/pygmt#3691]
**Release**: [v0.14.0](https://github.com/GenericMappingTools/pygmt/milestone/20)
**Scheduled Date**: 2024/12/31
**Pull request due date**: 2024/12/28
**DOI**: `10.5281/zenodo.14535921`

**Priority PRs/issues to complete prior to release**

- [x] #3683 
  - [x] #3685  @yvonnefroehlich 
  - [x] #3689  @yvonnefroehlich 
- [ ] ~~#3595~~ 
  - [ ] ~~#3598~~  @yvonnefroehlich  
  - [ ] ~~#3602~~  @seisman 
- [x] SPEC 0
  - [x] #3490 @weiji14
  - [x] #3606
  - [x] #3697
- [ ] GMTDataArrayAccessor
  - [x] #3693
  - [ ] ~~#3696~~ 
- [x] `Figure.hlines` #923
- [x] `Figure.vlines` #3726 
- [ ] Get rid of temporary files in grdcut and netCDF4
  - [ ] ~~#3115~~
  - [ ] ~~#3673~~
  - [ ] ~~#3643~~
- [ ] ~~#3105~~
- [x] Remote datasets
  - [x] #3706 @yvonnefroehlich 
  - [x] #3717 @yvonnefroehlich 
  - [x] #3718 @yvonnefroehlich 
  - [x] #3727 @yvonnefroehlich 
  - [x] #3728 @yvonnefroehlich 
  
**Before release**:

- [x] Check [SPEC 0](https://scientific-python.org/specs/spec-0000/) to see if we need to bump the minimum supported versions of GMT, Python and core package dependencies (NumPy, pandas, Xarray) @seisman
- [x] Review the ["PyGMT Team" page](https://www.pygmt.org/dev/team.html) @seisman
- [ ] Check to ensure that:
  - [x] Deprecations and related tests are removed for this version by running `grep --include="*.py" -r vX.Y.Z` from the base of the repository @seisman
  - [x] All tests pass in the ["GMT Legacy Tests" workflow](https://github.com/GenericMappingTools/pygmt/actions/workflows/ci_tests_legacy.yaml)
  - [ ] All tests pass in the ["GMT Dev Tests" workflow](https://github.com/GenericMappingTools/pygmt/actions/workflows/ci_tests_dev.yaml)
  - [x] All tests pass in the ["Doctests" workflow](https://github.com/GenericMappingTools/pygmt/actions/workflows/ci_doctests.yaml)
- [x] Update warnings in `pygmt/_show_versions.py` as well as notes in [Not working transparency](https://www.pygmt.org/dev/install.html#not-working-transparency) regarding GMT-Ghostscript incompatibility
- [X] Reserve a DOI on [Zenodo](https://zenodo.org) by clicking on "New Version" @seisman 
- [x] Finish up the "Changelog entry for v0.x.x" Pull Request (Use the previous changelog PR as a reference)
- [x] Run `make codespell` to check common misspellings. If there are any, either fix them or add them to `ignore-words-list` in `pyproject.toml`
- [x] Draft the announcement on https://hackmd.io/sxVhw5ReS3OVoeHgDmAxtA

**Release**:

- [x] At the [PyGMT release page on GitHub](https://github.com/GenericMappingTools/pygmt/releases):
  - [x] Edit the draft release notes with the finalized changelog
  - [x] Set the tag version and release title to vX.Y.Z
  - [x] Make a release by clicking the 'Publish Release' button, this will automatically create a tag too
- [x] Download pygmt-X.Y.Z.zip (rename to pygmt-vX.Y.Z.zip) and baseline-images.zip from
      the release page, and upload the two zip files to https://zenodo.org/deposit,
      ensure that they are filed under the correct reserved DOI

**After release**:

- [x] Update conda-forge [pygmt-feedstock](https://github.com/conda-forge/pygmt-feedstock) https://github.com/conda-forge/pygmt-feedstock/pull/38
      [Done automatically by conda-forge's bot. Remember to pin Python and SPEC0 versions]
- [x] Bump PyGMT version on https://github.com/GenericMappingTools/try-gmt (after conda-forge update) https://github.com/GenericMappingTools/try-gmt/pull/56
- [x] Announce the release on:
  - [x] GMT [forum](https://forum.generic-mapping-tools.org/c/news/) (do this announcement first! Requires moderator status)
  - [x] [ResearchGate](https://www.researchgate.net) (after forum announcement, add new version as research item via the **code** category, be sure to include the corresponding new Zenodo DOI)

---

- [x] Party :tada: (don't tick before all other checkboxes are ticked!)

**Comment by seisman** [comment]
@GenericMappingTools/pygmt-maintainers I think it's time to prepare for the v0.14.0 release. I've set the release date to 2024/12/31, but it's OK to postpone it to early Jan. 2025, depending on the amount of final changes that we want to make to the v0.14.0 release.

**Comment by yvonnefroehlich** [comment]
> @GenericMappingTools/pygmt-maintainers I think it's time to prepare for the v0.14.0 release. I've set the release date to 2024/12/31, but it's OK to postpone it to early Jan. 2025, depending on the amount of final changes that we want to make to the v0.14.0 release.

I have added some of my pending PR, but probably a bit ambious to finish them all for v0.14.0.

**Comment by seisman** [comment]
> I have added some of my pending PR, but probably a bit ambious to finish them all for v0.14.0.

PRs/Issues that we have set the v0.14.0 milestone can be tracked at https://github.com/GenericMappingTools/pygmt/milestone/20. We may finish them or not depending on our availability. In the release checklist, I think we should only list the high-priority PRs/issues that must be finished prior to the release.

**Comment by weiji14** [comment]
> > I have added some of my pending PR, but probably a bit ambious to finish them all for v0.14.0.
> > Would prefer to move the release not too much into January, as it will be likely that I will be only limited available for the first month / half of 2025.
> 
> PRs/Issues that we have set the v0.14.0 milestone can be tracked at https://github.com/GenericMappingTools/pygmt/milestone/20. We may finish them or not depending on our availability. In the release checklist, I think we should only list the high-priority PRs/issues that must be finished prior to the release.

We can keep the release to be end of December. @seisman, I feel that the PRs around making NetCDF4 optional and `pygmt.read` might need more time, would probably push that to the v0.15.0 release. The PR at #3115 can probably get in though, I should have time to review next week around Christmas break.

**Comment by seisman** [comment]
@GenericMappingTools/pygmt-maintainers 

I plan to merge PRs #3722, #3727, and #3726 tomorrow (2024/12/28) and then start working on the release changelog. From 12/28 to 12/31, documentation changes like #3720, and #3711 can still be merged, but no more other code changes (#2431 may be an exception depending on the progress).

**Comment by seisman** [comment]
I've created a draft release announcement at https://hackmd.io/sxVhw5ReS3OVoeHgDmAxtA.

**Comment by yvonnefroehlich** [comment]
> I've created a draft release announcement at https://hackmd.io/sxVhw5ReS3OVoeHgDmAxtA.

Thanks @seisman for starting to work on the draft release announcement, I made some additions and updates.

**Comment by seisman** [comment]
There is still one failure in the "GMT Dev Tests" workflow:
```python
________________________ test_sphinterpolate_no_outgrid ________________________

mars =      longitude  latitude  radius_m
0      278.231   -40.059      6888
1      302.360    15.487      -233
2       57.10...8.384   -32.260      1375
368    312.032    22.271     -4400
369    134.260    47.670     -3820

[370 rows x 3 columns]

    @pytest.mark.benchmark
    def test_sphinterpolate_no_outgrid(mars):
        """
        Test sphinterpolate with no set outgrid.
        """
        temp_grid = sphinterpolate(data=mars, spacing=1, region="g")
        assert temp_grid.dims == ("lat", "lon")
        assert temp_grid.gmt.gtype == 1  # Geographic grid
        assert temp_grid.gmt.registration == 0  # Gridline registration
        npt.assert_allclose(temp_grid.max(), 14628.144)
        npt.assert_allclose(temp_grid.min(), -6908.1987)
        npt.assert_allclose(temp_grid.median(), 118.96849)
>       npt.assert_allclose(temp_grid.mean(), 272.60593)
E       AssertionError: 
E       Not equal to tolerance rtol=1e-07, atol=0
E       
E       Mismatched elements: 1 / 1 (100%)
E       Max absolute difference among violations: 0.00024763
E       Max relative difference among violations: 9.0836965e-07
E        ACTUAL: array(272.60568, dtype=float32)
E        DESIRED: array(272.60593)
```
It's likely caused by xarray changes.

**Comment by seisman** [comment]
I've made the v0.14.0 release and everything works fine, but the "Publish to PyPI" workflow fails https://github.com/GenericMappingTools/pygmt/actions/runs/12561955854/job/35021691730:

```
Error: Attestation generation failure: /github/workspace/dist/pygmt-0.14.0.tar.gz already has a publish attestation: /github/workspace/dist/pygmt-0.14.0.tar.gz.publish.attestation
```
So v0.14.0 is not available on PyPI yet.

This upstream issue https://github.com/pypa/gh-action-pypi-publish/issues/283 explains the reason. In short, it's because the action (actually the `twine` tool called in this action) generates the attestation file (https://github.blog/news-insights/product-news/introducing-artifact-attestations-now-in-public-beta/), and it fails if a attestation file already exists. So, our CI job fails because we use the action twice, for TestPyPI and PyPI, respectively.

The issue also provides many solutions, e.g.,

- Set `attestations: false` https://github.com/pypa/gh-action-pypi-publish/issues/283#issuecomment-2499296440
- Use separate jobs for the TestPyPI and PyPI uploads: https://github.com/pypa/gh-action-pypi-publish/issues/283#issuecomment-2558477301
- Disable TestPyPI uploads for releases.

We need to decide what to do with the workflow, but before that we need to decide what to do with v0.14.0 release. We can either manually upload v0.14.0 to PyPI or have a v0.14.1 release after fixing the workflow.

**Comment by seisman** [comment]
> We need to decide what to do with the workflow, but before that we need to decide what to do with v0.14.0 release. We can either manually upload v0.14.0 to PyPI or have a v0.14.1 release after fixing the workflow.

I've decided to manually upload v0.14.0 following instructions at https://packaging.python.org/en/latest/guides/distributing-packages-using-setuptools/#packaging-your-project. Now v0.14.0 is on PyPI.

**Comment by yvonnefroehlich** [comment]
A happy and healthy year 2025 to everyone 🎆🎇!

Thanks @seisman for finishing the changelog and making the release. After roughly reading through the PyPI-upload issue I feel uploading v0.14.0 manually was the better, fast, and probably easier solution.

For me, the draft release announcement looks already quite good. But I am a bit unsure about the examples and conferences sections. Currently, the AGU24 contributions (preworkshop and talk) are mentioned within in the conference section at the end. But for other releases we included the addition to the “External resources” page in the section for the “Addition and updates on examples”. Should we mention the preworkshop twice or move both AGU24 contributions up to the examples section? What do others think?

**Comment by seisman** [comment]
I think the current draft announcement is good.

**Comment by michaelgrund** [comment]
Yes, announcement draft looks good!

**Comment by seisman** [comment]
@yvonnefroehlich Could you please make the announcements on the forum and ResearchGate?

**Comment by yvonnefroehlich** [comment]
> @yvonnefroehlich Could you please make the announcements on the forum and ResearchGate?

Thanks for the feedbacks @seisman and @michaelgrund! Just wanted to make the announcement on the forum, but currently I can’t access to GMT forum website; I will try it later again. Is there any planned maintenance?

**Comment by seisman** [comment]
I can access the forum, so the website is still on-line.

**Comment by michaelgrund** [comment]
Can also do the RG announcement later if @yvonnefroehlich has no time.

**Comment by seisman** [comment]
OK, I'll do the forum announcement in a few minutes.

**Comment by yvonnefroehlich** [comment]
> OK, I'll do the forum announcement in a few minutes.
> 
> Edit: Done at https://forum.generic-mapping-tools.org/t/pygmt-v0-14-0-released/5668.

OK. Thanks! 

> I can access the forum, so the website is still on-line.

I have still problems to access the forum (tried different browsers as well as my laptop and smartphone). It only works when using the VPN of my university. But good to hear, that the website is online.

**Comment by yvonnefroehlich** [comment]
After @michaelgrund did the RG upload, it looks like all box a ticked :tada:! Thanks everybody for their efforts for this release 🤗!
